platform/kobo/koreader.sh

```diff
@@ -117,19 +117,30 @@ if [ -e crash.log ]; then
 	mv -f crash.log.new crash.log
 fi
 
+# workaround 32-bit without KSM (KSM is indicated by ${ksmroot} being set)
+# shellcheck disable=2154
+if [ -z "${ksmroot+x:?}" ]; then
 # TODO: add support for 32bit framebuffer in koreader. This is a workaround
 # to run koreader in 16bpp. Useful since FW 4.0
+CUR_ROTATE="$(cat "/sys/class/graphics/fb0/rotate")"
 FB_BPP=$(cat /sys/class/graphics/fb0/bits_per_pixel)
-[ "$FB_BPP" -gt 16 ] && /usr/sbin/fbset -depth 16
-# fix rotation issues on Kobo Aura HD (dragon) and Kobo Aura H2O (dahlia)
+[ "${FB_BPP}" -gt 16 ] && /usr/sbin/fbset -depth 16
+# this suffices on normal devices, but H2O is stupid and sets the opposite
+echo "${CUR_ROTATE}" > "/sys/class/graphics/fb0/rotate"
+# this therefore turns it around on H2O (and other potential idiots) and merely
+# innocently repeats on sane devices
 # shellcheck disable=SC2094
-case "$PRODUCT" in "dragon" | "dahlia" ) cat /sys/class/graphics/fb0/rotate > /sys/class/graphics/fb0/rotate ;; esac
+cat "/sys/class/graphics/fb0/rotate" > "/sys/class/graphics/fb0/rotate"
```

Comment:
Correct, I believe this more likely to be future-proof unless a third deviant behavior is added to the mix. On those two this line results in the desired orientation (because they give you the opposite of what you want) while on other devices it's just harmless repetition.